libs/auth/src/inserted_smart_card_auth_with_memory_card.ts

```diff
@@ -74,11 +104,51 @@ export class InsertedSmartCardAuthWithMemoryCard
     this.updateAuthStatus(machineState, { type: 'check_pin', pin: input.pin });
   }
 
-  readCardData<T>(
+  async startCardlessVoterSession(
+    machineState: InsertedSmartCardAuthMachineState,
+    input: { ballotStyleId: BallotStyleId; precinctId: PrecinctId }
+  ): Promise<void> {
+    assert(this.config.allowedUserRoles.includes('cardless_voter'));
+
+    await this.checkCardReaderAndUpdateAuthStatus(machineState);
+    if (
+      this.authStatus.status !== 'logged_in' ||
+      this.authStatus.user.role !== 'poll_worker'
+    ) {
+      return;
+    }
```

Comment:
I'm landing on the following mental model for auth lib errors:

_For methods that transition auth status, e.g. `checkPin`, `logOut`, `startCardlessVoterSession`, `endCardlessVoterSession`_

- Use `assert` for states incompatible with the static config (`this.config`)
- For other incompatible states (e.g. trying to check PIN when you're not in the PIN-checking state or trying to start a cardless voter session when you're not logged in as a poll worker), just ignore the request and leave the auth status unchanged

_For card reading and writing methods, e.g. `programCard`, `unprogramCard`, `readCardData`, `writeCardData`_

- Use `assert` for states incompatible with the static config (`this.config`)
- Otherwise always use an error `Result` instead of throwing
- Methods that can perform access checks themselves should (e.g. `programCard` should check if you're logged in as a system administrator and return an error `Result` if need be). Methods that can't because the access checks are dependent on the context in which they're used (e.g. `readCardData`) should leave those checks to backend consumers. Those backend consumers should also use an error `Result` instead of throwing.

You'll see changes in this and the next few commits to align with the above!

Reply:
Thanks for outlining this mental model. A couple of questions:
- What's the motivation for "just ignore the request and leave the auth status unchanged" vs failing more loudly?
- Why do access checks in methods like `programCard` return an error `Result`? To me, that would imply that the consumer isn't responsible for checking the auth before calling the method and needs to explicitly think about an "unauthorized" use case (e.g. if we allowed any user type to press a "program card" button ). Vs using an `assert` would imply that the system should only allow users to call `programCard` if they're logged in as a sys admin and if that doesn't happen, the programmer made a mistake.

Reply:
Yes good questions!

The motivation for both of the above is actually similar. I've been thinking about the case that someone triggers an action in a valid state but then does something like removing their card right after, switching to an invalid state. I wouldn't want to trigger the error boundary in that case.

I know this may not be likely in practice since the timing would have to be just right, but the fact that all these methods depend not only on code but the status of the card in the card reader makes me lean toward treating errors as user errors instead of code errors that trigger the error boundary.

For what it's worth, I have been able to trigger the error boundary by updating the `checkPin` method to assert that `this.authStatus.status === 'checking_passcode'` and removing my card right after entering the last digit of my PIN:

https://user-images.githubusercontent.com/12616928/218866133-7549f67f-8e68-4ffc-94bf-a5f792ca02b3.mov

Reply:
Ok cool that makes sense. Maybe leave some documentation of this approach somewhere so we preserve it over time? Seems like something that could be easily forgotten

Reply:
Added a README and documented there: https://github.com/votingworks/vxsuite/pull/3011/commits/b7f9a9aa72033030590c7f7566225d3c9a925750